// everscale-standalone-client/client/AccountsStorage/Generic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsigAccount = exports.GenericAccount = void 0;
const everscale_inpage_provider_1 = require("everscale-inpage-provider");
/**
 * @category AccountsStorage
 */
class GenericAccount {
    constructor(args) {
        this.address = args.address instanceof everscale_inpage_provider_1.Address ? args.address : new everscale_inpage_provider_1.Address(args.address);
        this.abi = typeof args.abi === 'string' ? args.abi : JSON.stringify(args.abi);
        this.prepareMessageImpl = args.prepareMessage;
        this.publicKey = args.publicKey;
    }
    async fetchPublicKey(ctx) {
        if (this.publicKey != null) {
            return this.publicKey;
        }
        this.publicKey = await ctx.fetchPublicKey(this.address);
        return this.publicKey;
    }
    async prepareMessage(args, ctx) {
        const publicKey = await this.fetchPublicKey(ctx);
        const signer = await ctx.getSigner(publicKey);
        const { method, params, stateInit } = await this.prepareMessageImpl(args, ctx);
        return await ctx.createExternalMessage({
            address: this.address,
            signer,
            timeout: args.timeout,
            abi: this.abi,
            method,
            params,
            stateInit,
        });
    }
}
exports.GenericAccount = GenericAccount;
/**
 * @category AccountsStorage
 */
class MsigAccount extends GenericAccount {
    constructor(args) {
        super({
            address: args.address,
            publicKey: args.publicKey,
            abi: MSIG_ABI,
            prepareMessage: async (args, ctx) => {
                const payload = args.payload
                    ? ctx.encodeInternalInput(args.payload)
                    : '';
                return {
                    method: 'sendTransaction',
                    params: {
                        dest: args.recipient,
                        value: args.amount,
                        bounce: args.bounce,
                        flags: 3,
                        payload,
                    },
                };
            },
        });
    }
}
exports.MsigAccount = MsigAccount;
const MSIG_ABI = `{
  "ABI version": 2,
  "header": ["pubkey", "time", "expire"],
  "functions": [{
    "name": "sendTransaction",
    "inputs": [
      {"name":"dest","type":"address"},
      {"name":"value","type":"uint128"},
      {"name":"bounce","type":"bool"},
      {"name":"flags","type":"uint8"},
      {"name":"payload","type":"cell"}
    ],
    "outputs": []
  }],
  "events": []
}`;
